### ISSUE TITLE
Broken links

### ISSUE THREAD

**Issue by svick** [dotnet/docs#1815]
While trying to fix broken links (https://github.com/dotnet/docs/pull/1814), I encountered several that I couldn't fix. Either their targets should be fixed, or the links should be removed.

* [Additional Resources for Visual Basic Programmers](https://docs.microsoft.com/en-us/dotnet/articles/visual-basic/getting-started/additional-resources) links to http://msdn.microsoft.com/practices/default.aspx, which redirects to http://pnp.azurewebsites.net/, which errors out with "The specified CGI application encountered an error and the server terminated the process."
* Two articles contain links to articles in https://docs.microsoft.com/visualstudio/data-tools/, which were deleted in https://github.com/Microsoft/visualstudio-docs/commit/c2afa3e2f52109eb01e47f1731a19f9f74810c6d. It seems these articles only exist in old versions of MSDN docs, but don't have any current version. Specifically:

   *  [Accessing data in Visual Basic applications](https://docs.microsoft.com/en-us/dotnet/articles/visual-basic/developing-apps/accessing-data) links to:
      * creating-data-applications
      * overview-of-data-applications-in-visual-studio
      * connecting-to-data-in-visual-studio
      * fetching-data-into-your-application
      * editing-data-in-your-application
   * [Nullable Value Types (Visual Basic)](https://docs.microsoft.com/en-us/dotnet/articles/visual-basic/programming-guide/language-features/data-types/nullable-value-types) links to tableadapter-overview
* [COM Interoperability in .NET Framework Applications (Visual Basic)](https://docs.microsoft.com/en-us/dotnet/articles/visual-basic/programming-guide/com-interop/com-interoperability-in-net-framework-applications) links to https://docs.microsoft.com/cpp/misc/com-interop-wrapper-error, which was deleted in https://github.com/Microsoft/cpp-docs/commit/3fa92896d ([the article before deletion](https://github.com/Microsoft/cpp-docs/blob/87824e5e183204806cd95c362c1c2bf0998dae68/docs/misc/com-interop-wrapper-error.md)).
* [How to: Call a Web Service Asynchronously (Visual Basic)](https://docs.microsoft.com/en-us/dotnet/articles/visual-basic/developing-apps/programming/how-to-call-a-web-service-asynchronously) links to http://www.xmethods.net/, which seems to be down.

**Comment by GuardRex** [comment]
Side Note: I never got around to it over in the ASP.NET Core docs, but there were quite a few 301 redirects over there (e.g., `http://twitter.com` should be `https://twitter.com`). I haven't done the analysis over here for redirects. It's a separate issue perhaps, but I was just wondering if when you did your broken link analysis if you witnessed a lot of 301's as well?

**Comment by svick** [comment]
@GuardRex Looking at it now, there seems to be a lot of 301s, starting with 23 of them just on [the main page](https://docs.microsoft.com/en-us/dotnet/). And the vast majority of those are due to a missing trailing slash (e.g. https://docs.microsoft.com/en-us/dotnet/articles/core is 301 pointing to https://docs.microsoft.com/en-us/dotnet/articles/core/). Personally I don't think those are worth fixing.

Though I guess other redirects, like https://azure.microsoft.com/documentation/services/storage/, which is 301 pointing to https://docs.microsoft.com/en-us/azure/storage/ (with a 302 detour to https://azure.microsoft.com/en-us/documentation/services/storage/) might be worth it.

I haven't looked at redirects any further though.

In case you're interested, [the code I used is here](https://github.com/svick/DotnetCoreDocsWalker/blob/dotnet-docs-only/Program.cs).

**Comment by GuardRex** [comment]
> Personally I don't think those are worth fixing.

*Roundtrips suck!* lol :grinning: I'd fix all of them. I'm a bit of speed freak, so I don't like anything that slows anything down.

> DotnetCoreDocsWalker

Ah ... that's very interesting. I use [Xanu](http://home.snafu.de/tilman/xenulink.html) ... must be their :alien: influences :smile:, but I look forward to taking a closer look at your `DotnetCoreDocsWalker`.

@mairaw If you'd like an issue to fix 301's, let me know and I'll open one for it.

**Comment by mairaw** [comment]
@GuardRex can't we use this issue for that? I personally prefer going straight to the topic avoiding 301 redirects, if possible.

@mikeblome are you guys deleting content without establishing 301 redirects for them in the VS docs?

**Comment by GuardRex** [comment]
Sure. I just didn't know if it was something you wanted to table for a later date.

Yes, indeed on 301's ... eee ... yuck! :grinning: ... and they have quite a few over in ASP.NET Core docs IIRC. I'll check both sets of docs and talk to Tom about the ones over there and report back here what I find out.

**Comment by mairaw** [comment]
Yes, it could go to the backlog. I recently had to add all the pages where we missed the 301 redirects in our repo. They do run reports and open bugs for those internally and also on 404's (although I haven't seen a recent version of a 404 report).

**Comment by mikeblome** [comment]
I'm not aware of any deletions. +Mike and Kemp


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by mairaw** [comment]
@mikeblome this is the commit that @svick pointed to in the VS repo: https://github.com/Microsoft/visualstudio-docs/commit/c2afa3e2f52109eb01e47f1731a19f9f74810c6d

This broken a bunch of our links to the VS docs.

**Comment by mikeblome** [comment]
ah ok. darn. These topics were removed in CAPS last year, but we stashed them in NOT in TOC rather than Retired Contents as a hack to get around some CAPS issue. So they got migrated to docs. I'll create redirects.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by GuardRex** [comment]
Took a look. I see a bunch. To get an accurate count, I'd need to factor out the 301's for localization. However, a bunch jump right out (this is just a sample) ...

https://docs.microsoft.com/en-us/dotnet/articles/core/
http://asp.net/ :point_right: https://asp.net/
	
https://docs.microsoft.com/en-us/dotnet/articles/core/macos-prerequisites
http://brew.sh/ :point_right: https://brew.sh/

https://docs.microsoft.com/en-us/dotnet/articles/core/
http://creativecommons.org/licenses/by/4.0/ :point_right: https://creativecommons.org/licenses/by/4.0/

https://docs.microsoft.com/en-us/dotnet/articles/core/	
http://dotnetfoundation.org/ :point_right: https://dotnetfoundation.org/

http://www.dotnetfoundation.org/blog/tsg-welcome
http://dotnetfoundation.org/blog/tsg-welcome :point_right: https://dotnetfoundation.org/blog/tsg-welcome

https://docs.microsoft.com/en-us/dotnet/articles/core/tutorials/using-on-macos
http://nuget.org/ :point_right: http://www.nuget.org/

Here's one in an author byline for Mr. Nate ...

https://docs.microsoft.com/en-us/dotnet/articles/core/tools/project-json-to-csproj
http://github.com/natemcmaster :point_right: https://github.com/natemcmaster

These look systemically broken ...

https://docs.microsoft.com/en-us/dotnet/articles/core/packages
http://docs.microsoft.com/dotnet/core/api/System.Reflection.Assembly :point_right: https://docs.microsoft.com/dotnet/core/api/System.Reflection.Assembly

... and it looks like there are dozens of those redirecting to the secure endpoint for the API docs.

**Comment by mairaw** [comment]
I've opened PR #14192 to fix the redirections on the homepage.

I think most of the other redirections and broken links were already fixed throughout the years.
The only exception I found is the last bullet:
* How to: Call a Web Service Asynchronously (Visual Basic) links to http://www.xmethods.net/, which seems to be down. 

But I'd prefer if we open a separate issue for that alone so that we can decide what to do with that article. It requires an update rather than a simple link fix.

Sounds good @svick?
